[LBSE] Fix RenderSVGResourceClipper (re)painting problems

rwlbuis opened the pull request:
#### e98de6f4b331de2ca6ea78c00b6b6d6243c78621
<pre>
[LBSE] Fix RenderSVGResourceClipper (re)painting problems
<a href="https://bugs.webkit.org/show_bug.cgi?id=264946">https://bugs.webkit.org/show_bug.cgi?id=264946</a>

Reviewed by Nikolas Zimmermann.

The logic in RenderSVGResourceClipper::resourceBoundingBox does not take
the actual clip content rectangle into account, use the legacy solution which does.

Also stop repainting SVG resources by adjusting RenderElement::isInsideEntirelyHiddenLayer.

This PR also adds repaint tests that verify repaint rectangles after changing clipping content.

* LayoutTests/svg/repaint/clip-path-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-shrink.html: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-shrink.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::isInsideEntirelyHiddenLayer const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::enclosingSVGHiddenOrResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::calculateClipContentRepaintRect):
* Source/WebCore/svg/SVGClipPathElement.h:

Canonical link: <a href="https://commits.webkit.org/270913@main">https://commits.webkit.org/270913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbe802a7377512f44addbe4d260354334518cd36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26736 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28948 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27181 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2747 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24356 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26998 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4194 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22949 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3666 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3726 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23941 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29433 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24408 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24341 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29969 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1935 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27870 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5192 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23677 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6440 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4212 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4095 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -211,27 +233,25 @@ bool RenderSVGResourceClipper::hitTestClipContent(const FloatRect& objectBoundin
-
-    if (layer()->isTransformed())
-        clipRect = layer()->currentTransform(RenderStyle::individualTransformOperations()).mapRect(clipRect);
+    if (m_clipBoundaries[repaintRectCalculation].isEmpty())
```
Oh, the caching logic no longer works for LBSE, after my resource invalidation patch, you cannot rely on removeAllClientsFrom...

I would not cache m_clipBoundaries at all, just compute it.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.h:
```diff
@@ -60,6 +60,10 @@ class RenderSVGResourceClipper final : public RenderSVGResourceContainer {
 
+    void calculateClipContentRepaintRect(RepaintRectCalculation);
+
+    EnumeratedArray<RepaintRectCalculation, FloatRect, RepaintRectCalculation::Accurate> m_clipBoundaries;
```
TIL EnumeratedArray :-)

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +196,23 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
     frameView.setPaintBehavior(oldBehavior);
 }
 
+void RenderSVGResourceClipper::calculateClipContentRepaintRect(RepaintRectCalculation repaintRectCalculation)
```
That should be moved into SVGClipPathElement, just like shouldApplyPathClipping, no?

rwlbuis replied:
Sorry, missed this one, will fix.

nikolaszimmermann reviewed (changes requested):
Almost there...

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +191,24 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
+{
+    FloatRect clipContentRepaintRect;
+    // This is a rough heuristic to appraise the clip size and doesn't consider clip on clip.
+    for (Node* childNode = clipPathElement().firstChild(); childNode; childNode = childNode->nextSibling()) {
```
This could need some autos...

rwlbuis replied:
Sure, done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +191,24 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
+        const RenderStyle& style = renderer->style();
+        if (style.display() == DisplayType::None || style.visibility() != Visibility::Visible)
+            continue;
+        clipContentRepaintRect.unite(renderer->localToParentTransform().mapRect(renderer->repaintRectInLocalCoordinates(repaintRectCalculation)));
```
There is no localToParentTransform in LBSE.

rwlbuis replied:
Good point, removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +191,24 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
+            continue;
+        clipContentRepaintRect.unite(renderer->localToParentTransform().mapRect(renderer->repaintRectInLocalCoordinates(repaintRectCalculation)));
+    }
+    return clipPathElement().animatedLocalTransform().mapRect(clipContentRepaintRect);
```
This also is not correct in LBSE, transformation handling works differently.

rwlbuis replied:
Fixed.

nikolaszimmermann reviewed (approved):
Hmm. still needs some work

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +191,24 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
+        auto* renderer = childNode->renderer();
+        if (!childNode->isSVGElement() || !renderer)
+            continue;
+        if (!renderer->isRenderOrLegacyRenderSVGShape() && !renderer->isRenderSVGText() && !childNode->hasTagName(SVGNames::useTag))
```
We shouldn’t be able to get here with a legacy render SVG shape..

rwlbuis replied:
Well spotted, will fix.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -191,6 +191,24 @@ void RenderSVGResourceClipper::applyMaskClipping(PaintInfo& paintInfo, const Ren
+        auto& style = renderer->style();
+        if (style.display() == DisplayType::None || style.visibility() != Visibility::Visible)
+            continue;
+        clipContentRepaintRect.unite(renderer->repaintRectInLocalCoordinates(repaintRectCalculation));
```
Need to check, if the repaint rect accessor usage  is correct.

nikolaszimmermann replied:
Seems fine - this includes the transformations of the children, which is what we need here.

nikolaszimmermann reviewed (commented):
We should add some more repaint tests that really check if things are working - we need dumps of the target repaint rect if the clipPath has regular or transformed kids, in both userSpaceOnUse and objectBoundingbox mode.

I think I recently added some tests to svg/repaint that could be used for starting. I would like to make sure that these fundamentals are covered with tests.

nikolaszimmermann reviewed (approved):
Looks good - except the repaint rects, as discussed internally, they are still wrong, but at least documented in the new tests. We need to land this as prerequisite to rework the SVG resources in LBSE, where we will also visit the repaint rect topic again, to make sure it's 100% correct in future (no more spurious repaint rects, as small as necessary)

rwlbuis committed:
[LBSE] Fix RenderSVGResourceClipper (re)painting problems
https://bugs.webkit.org/show_bug.cgi?id=264946

Reviewed by Nikolas Zimmermann.

The logic in RenderSVGResourceClipper::resourceBoundingBox does not take
the actual clip content rectangle into account, use the legacy solution which does.

Also stop repainting SVG resources by adjusting RenderElement::isInsideEntirelyHiddenLayer.

This PR also adds repaint tests that verify repaint rectangles after changing clipping content.

* LayoutTests/svg/repaint/clip-path-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-shrink.html: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-shrink.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/clip-path-user-space-on-use.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::isInsideEntirelyHiddenLayer const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::enclosingSVGHiddenOrResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::calculateClipContentRepaintRect):
* Source/WebCore/svg/SVGClipPathElement.h:

Canonical link: https://commits.webkit.org/270913@main
